### PR TITLE
Misc fixes + minimum macro support

### PR DESCRIPTION
**Pull request by yamt**

### CODE REVIEW

**Comment by mayanez** [comment]
Oh looks super cool! Thanks for the PR :) Nice to see you figured out the macro stuff. I'll take a look this weekend before I merge it in. What OS did you test this on?

**Comment by yamt** [comment]
> Oh looks super cool! Thanks for the PR :) Nice to see you figured out the macro stuff. I'll take a look this weekend before I merge it in. What OS did you test this on?

macOS.

File: dumang_ctrl/dumang/common.py [hunk 1]
```diff
@@ -621,11 +685,51 @@ def __init__(self, key, layer_keycodes):
         self.layer_keycodes = {k: v.encode() if isinstance(v, Keycode) else v for k, v in layer_keycodes.items()}
 
     def encode(self):
-        return [self.cmd, 0x01, self.layer_keycodes[1], self.layer_keycodes[2], self.layer_keycodes[3], 0xFF, self.layer_keycodes[0]]
+        return [self.cmd, self.key.encode() + 1, self.layer_keycodes[1], self.layer_keycodes[2], self.layer_keycodes[3], 0xFF, self.layer_keycodes[0]]
```
Comment (mayanez) [inline dumang_ctrl/dumang/common.py:688]: Can you elaborate on this a bit? What's this byte actually for? I wasn't able to decipher it the first time around. Also why the `+ 1`?

File: dumang_ctrl/dumang/common.py [hunk 2]
```diff
@@ -481,22 +536,29 @@ def configured_key(self, k):
...
+                    if any([kc.keycode != 0 for kc in p.layer_keycodes.values()]):
+                        # NOTE: We add the layer_keycodes to the DKM
+                        p.key.layer_keycodes = p.layer_keycodes
+                        self._configured_keys[p.key.key] = DuMangKeyModule(p.key, p.layer_keycodes)
+                    if any([kc.keycode == Keycode.MACRO for kc in p.layer_keycodes.values()]):
+                        self.put(MacroReportRequestPacket(p.key, 0))
+                        pending += 1
+                if isinstance(p, MacroReportResponsePacket):
+                    pending -= 1
+                    if p.type.type not in [0, 0xff]:
```
Comment (mayanez) [inline dumang_ctrl/dumang/common.py:557]: Which `type` is this referring to? Just trying to get some understanding on the `0` to `0xff`.

**Review by mayanez — COMMENTED** [review]
Sorry for the delay. Thanks again for this change, looks great!

I had a few questions on some things I was unsure of before I go ahead and merge it in.

File: dumang_ctrl/dumang/common.py (hunk 1 above)
Comment (yamt) [reply dumang_ctrl/dumang/common.py:688]: it's just my attempt to mimic what the official app does.

File: dumang_ctrl/dumang/common.py (hunk 2 above)
Comment (yamt) [reply dumang_ctrl/dumang/common.py:557]: it's MacroType.type.
while i don't really know what 0 and 0xff are,
they seemed like a termination signal of a series of packets.

**Pull request merged**